File: CSE3063F22_iteration2_classes/Transcript.java
import java.util.ArrayList;
import java.util.List;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;


public class Transcript {

    private String studentId;
    private List<Course> passedCourses;
    private List<Course> Schedule;
    private List<Grade> grades;

    public Transcript(String studentId) throws FileNotFoundException, IOException, ParseException {
        this.studentId = studentId;
        Object obj = new JSONParser().parse(new FileReader("students/"+getStudentId()+".json"));
        JSONObject jo = (JSONObject) obj;

        JSONArray PassedCourses = (JSONArray) jo.get("PassedCourses");
        this.passedCourses = new ArrayList<>();

        setpassedCourses(PassedCourses);

        JSONArray Schedule = (JSONArray) jo.get("Schedule");
        this.Schedule = new ArrayList<>();
        setSchedule(Schedule);
        
        this.grades = new ArrayList<>();
    }


    public String getStudentId() {
        return studentId;
    }

    public List<Course> getpassedCourses() {
        return passedCourses;
    }

    public void setpassedCourses(List<Course> courses) {
        this.passedCourses = courses;
    }

    public List<Course> getSchedule() {
        return Schedule;
    }

    public void setSchedule(List<Course> courses) {
        this.Schedule = courses;
    }

    public void addCoursetoSchedule(Course course) {
        this.Schedule.add(course);
        System.out.println("yeni schedule: "+getSchedule().toString());

    }

    
    public boolean GPAchecked() throws FileNotFoundException, IOException, ParseException{
        boolean isIt = true;
        Object obj = new JSONParser().parse(new FileReader("students/"+getStudentId()+".json"));
        JSONObject jo = (JSONObject) obj;
        String GPA = (String) jo.get("GPA");
        double gpa =Double.parseDouble(GPA);


        if( gpa < 2.00){
            isIt=false;
            System.out.println("Below 2 GPA");
            

        }
        else{
            isIt=true;
            System.out.println("Above 2 GPA");
            
        }

        return isIt;


    }
    
    public boolean isCourseNotPassedBefore(Course course){
        boolean isIt=true;
        List<Course> student_passed = getSchedule();
        List<Course> student_transcript = getpassedCourses();
        String[] passed_before = new String[student_passed.size()];
        String[] transcript = new String[student_transcript.size()];

        if((student_transcript.size()==0)){
            isIt=true;
        }
    
        else{
    
        for(int i=0;i<student_transcript.size();i++){
        
      
        transcript[i] = String.valueOf(student_transcript.get(i));
    
        }
    
        int i=0;

        while(i<transcript.length){
        
        if((transcript[i].equals(course.getCourseId()))==true){
            isIt=false;
        }
        i++;
        }
    

        for(int x=0;x<student_passed.size();x++){

            passed_before[x]=String.valueOf(student_passed.get(x));
    
            }
    
            int x=0;
            while(x<passed_before.length){
              
            if((passed_before[x].equals(course.getCourseId()))==true){
                isIt=false;
            }
            x++;
            }
        }

    return isIt;

    }
    
}

File: ITERATION3/OOSD_3RD_ITERATİON_code/Transcript.py
import json

from Course import Course
class Transcript(object):
    def __init__(self, student_id):
        self.__student_id = student_id
        try:
         with open(f"students/{self.__student_id}.json", "r") as f:
            data = json.load(f)
        
            self.__passed_courses = []
            for course in data["PassedCourses"]:
                self.__passed_courses.append(Course(course).get_course_id())

            self.__Schedule = []
            for course in data["Schedule"]:
                self.__Schedule.append(Course(course).get_course_id())

            self.__grades = []
        except FileNotFoundError:
            print('No file with this student id for Trnscript')

    def get_student_id(self):
        return self.__student_id

    def get_passed_courses(self):
        return self.__passed_courses

    def set_passed_courses(self, courses):
        self.__passedCourses = courses

    def get_schedule(self):
        return self.__Schedule

    def set_schedule(self, courses):
        self.__Schedule = courses

    def add_course_to_schedule(self, course : Course):
        self.__Schedule.append(course.get_course_id())

    def GPA_checked(self):
        with open(f"students/{self.__student_id}.json") as file:
            data = json.load(file)
        gpa = data["GPA"]
        if gpa < 2.0:
            return False
        else:
            return True
    
    def is_course_not_passed_before(self,course:Course):
        if ((course.get_course_id() in self.get_passed_courses()) | (course.get_course_id() in self.get_schedule())):
            return False
        else:
            return True
